src/context.rs

```diff
@@ -65,6 +70,23 @@ impl<'ctx> MessageContext<'ctx> {
         self.terms.get(id)
     }
 
+    pub fn add_function<F>(&mut self, id: &str, func: F)
+    where
+        F: 'static
```

Comment:
is there a reason for using `'static` rather than `'ctx`?

Reply:
I was unable to get `'ctx` to compile

Reply:
If it helps, the following changes worked for me:

```diff
diff --git a/src/context.rs b/src/context.rs
index 32ab07a..1caaf69 100644
--- a/src/context.rs
+++ b/src/context.rs
@@ -39,17 +39,19 @@ use super::resolve::{Env, ResolveValue};
 /// to negotiate a sensible fallback for date and time formatting.
 #[allow(dead_code)]
 pub struct MessageContext<'ctx> {
     pub locales: &'ctx [&'ctx str],
     messages: HashMap<String, ast::Message>,
     terms: HashMap<String, ast::Term>,
     functions: HashMap<
         String,
-        Box<Fn(&[Option<FluentValue>], &HashMap<String, FluentValue>) -> Option<FluentValue>>,
+        Box<
+            'ctx + Fn(&[Option<FluentValue>], &HashMap<String, FluentValue>) -> Option<FluentValue>,
+        >,
     >,
 }
 
 impl<'ctx> MessageContext<'ctx> {
     pub fn new(locales: &'ctx [&'ctx str]) -> MessageContext {
         MessageContext {
             locales,
             messages: HashMap::new(),
@@ -67,27 +69,28 @@ impl<'ctx> MessageContext<'ctx> {
     }
 
     pub fn get_term(&self, id: &str) -> Option<&ast::Term> {
         self.terms.get(id)
     }
 
     pub fn add_function<F>(&mut self, id: &str, func: F)
     where
-        F: 'static
-            + Fn(&[Option<FluentValue>], &HashMap<String, FluentValue>) -> Option<FluentValue>,
+        F: 'ctx + Fn(&[Option<FluentValue>], &HashMap<String, FluentValue>) -> Option<FluentValue>,
     {
         self.functions.insert(id.to_string(), Box::new(func));
     }
 
     pub fn get_function(
         &self,
         id: &str,
     ) -> Option<
-        &Box<Fn(&[Option<FluentValue>], &HashMap<String, FluentValue>) -> Option<FluentValue>>,
+        &Box<
+            'ctx + Fn(&[Option<FluentValue>], &HashMap<String, FluentValue>) -> Option<FluentValue>,
+        >,
     > {
         self.functions.get(id)
     }
 
     pub fn add_messages(&mut self, source: &str) {
         let res = parse(source).unwrap_or_else(|x| x.0);
 
         for entry in res.body {

```